File: apps/api/services/healer.py
import datetime
import json
import logging
import os
from typing import Dict, List
from sqlalchemy.orm import Session
from database.models import Scraper, Run, HealingAttempt
from .brightdata import BrightDataService
from .detector import FailureDetector
from .validator import OutputValidator

logger = logging.getLogger("scrapeguard.healer")

MAX_HEAL_ATTEMPTS = 2

class HealerService:
    @classmethod
    def calculate_confidence_score(cls, validation_result: Dict = None, historical_avg: float = None) -> int:
        """
        Calculates a repair confidence score (0-100%) based primarily on post-repair validation.
        - 90-100%: Automatic recovery accepted.
        - 70-89%: Recovery accepted but flagged for review.
        - <70%: Escalate to human review.
        """
        if not validation_result or not validation_result.get("success", False):
            return 30  # Default failure score

        score = 70  # Base score for passing validation
        
        summary = validation_result.get("summary", {})
        fields_presence = summary.get("fields_presence", {})
        fields_type_valid = summary.get("fields_type_valid", {})
        total_records = summary.get("total_records", 0)

        # 1. Field completeness bonus (up to 10%)
        # If all fields have 100% presence, add 10%
        if fields_presence:
            avg_presence = sum(fields_presence.values()) / len(fields_presence)
            score += int(avg_presence * 10)

        # 2. Schema validity bonus (up to 10%)
        # If all types are 100% valid, add 10%
        if fields_type_valid:
            avg_type_valid = sum(fields_type_valid.values()) / len(fields_type_valid)
            score += int(avg_type_valid * 10)

        # 3. Historical count recovery bonus (up to 10%)
        if historical_avg is not None and historical_avg > 0:
            count_ratio = min(total_records, historical_avg) / max(total_records, historical_avg)
            score += int(count_ratio * 10)
        else:
            # If no history, give a static 5% bonus for having records
            if total_records > 0:
                score += 5

        return min(max(score, 0), 100)

    @classmethod
    def trigger_healing_process(
        cls,
        db: Session,
        scraper_id: str,
        failed_run_id: str
    ) -> HealingAttempt:
        """
        Orchestrates the healing lifecycle.
        - Generates failure diagnostics
        - Triggers Bright Data healing
        - Reruns scraper collector
        - Validates new output
        - Updates DB states
        """
        scraper = db.query(Scraper).filter(Scraper.id == scraper_id).first()
        failed_run = db.query(Run).filter(Run.id == failed_run_id).first()

        if not scraper or not failed_run:
            logger.error(f"Cannot heal: Scraper {scraper_id} or Run {failed_run_id} not found.")
            raise ValueError("Scraper or Run not found")

        # 1. Fetch previous successful runs to establish average record count baseline
        success_runs = db.query(Run).filter(
            Run.scraper_id == scraper_id,
            Run.status == "SUCCESS"
        ).order_by(Run.completed_at.desc()).limit(5).all()
        
        hist_avg = None
        if success_runs:
            hist_avg = sum(r.records_count for r in success_runs) / len(success_runs)

        # 2. Perform failure detection on the failed run's results
        # We need the raw results from the failed run. Since it failed, validation_status is populated
        val_status = failed_run.validation_status or {}
        
        failure_description = FailureDetector.generate_diagnostic_summary(
            scraper.name,
            scraper.collector_id or "c_mock",
            scraper.schema_definition,
            val_status
        )

        # 3. Record Healing Attempt
        attempt = HealingAttempt(
            scraper_id=scraper.id,
            run_id=failed_run.id,
            status="HEALING",
            failure_description=failure_description,
            collector_id=scraper.collector_id,
            records_before=failed_run.records_count,
            started_at=datetime.datetime.utcnow()
        )
        db.add(attempt)
        failed_run.status = "HEALING"
        scraper.status = "HEALING"
        db.commit()

        # 3.5. Try Local Selector Version Recovery first
        from database.models import SelectorVersion
        # Fetch previous versions, ordering by success_count desc (most successful first)
        historical_versions = db.query(SelectorVersion).filter(
            SelectorVersion.scraper_id == scraper_id
        ).order_by(SelectorVersion.success_count.desc(), SelectorVersion.version.desc()).all()
        
        local_healed = False
        healed_selectors = None
        healed_version_obj = None
        test_run_res = None
        test_val_res = None
        
        # Keep track of active selectors so we can restore if all else fails
        original_selectors = dict(BrightDataService.get_selectors(scraper_id=scraper.id))

        for hv in historical_versions:
            # Skip if they match current active selectors
            if hv.selectors == original_selectors:
                continue
                
            logger.info(f"Attempting local selector recovery using version {hv.version} for scraper {scraper_id}...")
            # Temporarily apply historical selectors
            BrightDataService.set_selectors(hv.selectors, scraper_id=scraper.id)
            
            # Execute test run using this selector set (bypassing pre-flight cache)
            test_run_res = BrightDataService.run_collector(
                collector_id=scraper.collector_id,
                target_url=scraper.target_url,
                schema=scraper.schema_definition,
                scraper_id=scraper.id,
                db=db,
                bypass_cache=True
            )
            
            if test_run_res["status"] == "SUCCESS":
                # Validate the results of this trial
                test_val_res = OutputValidator.validate_records(
                    records=test_run_res["records"],
                    schema=scraper.schema_definition,
                    expected_min_count=1,
                    historical_avg_count=int(hist_avg) if hist_avg else None
                )
                if test_val_res["success"]:
                    logger.info(f"Local selector recovery SUCCESS using version {hv.version}!")
                    local_healed = True
                    healed_selectors = hv.selectors
                    healed_version_obj = hv
                    hv.success_count += 1
                    break
                    
        if local_healed and test_run_res and test_val_res and healed_version_obj:
            # Found working historical selectors! Update db states
            attempt.status = "SUCCESS"
            attempt.completed_at = datetime.datetime.utcnow()
            attempt.error = None
            attempt.records_after = len(test_run_res["records"])
            attempt.validation_result = test_val_res
            
            failed_run.status = "HEALED"
            failed_run.completed_at = datetime.datetime.utcnow()
            failed_run.records_count = len(test_run_res["records"])
            failed_run.validation_status = test_val_res
            failed_run.recovery_source = "LOCAL_VERSION_RECOVERY"
            failed_run.duration = (failed_run.completed_at - failed_run.started_at).total_seconds()
            
            result_filename = f"run_{failed_run.id}_healed.json"
            cls._save_raw_result(result_filename, test_run_res["records"])
            failed_run.raw_result_reference = result_filename
            
            scraper.status = "HEALTHY"
            scraper.current_version = healed_version_obj.version
            scraper.last_success = datetime.datetime.utcnow()
            scraper.last_run = datetime.datetime.utcnow()
            
            # Cache the successful run
            BrightDataService.update_scraper_cache(db, scraper.id, test_run_res["records"], test_run_res.get("hash"))
            
            # Perform cross-scraper correlation
            cls.correlate_scrapers_selector_update(db, scraper_id, healed_selectors)
            
            cls._update_scraper_success_rate(db, scraper_id)
            db.commit()
            return attempt

        # Restore original selectors if local healing failed
        BrightDataService.set_selectors(original_selectors, scraper_id=scraper.id)

        # 4. Invoke Bright Data CLI / API to heal (Fallback with retry)
        last_error = None
        new_val_res = None

        for heal_attempt in range(1, MAX_HEAL_ATTEMPTS + 1):
            if heal_attempt > 1:
                logger.info(f"Retrying AI heal (attempt {heal_attempt}/{MAX_HEAL_ATTEMPTS})...")

            heal_res = BrightDataService.heal_collector(
                collector_id=scraper.collector_id,
                failure_description=failure_description,
                scraper_id=scraper.id
            )

            if heal_res["status"] != "SUCCESS":
                last_error = f"Bright Data CLI Healing failed: {heal_res['error']}"
                continue

            attempt.status = "VALIDATING"
            db.commit()

            run_res = BrightDataService.run_collector(
                collector_id=scraper.collector_id,
                target_url=scraper.target_url,
                schema=scraper.schema_definition,
                scraper_id=scraper.id,
                db=db,
                bypass_cache=True
            )

            if run_res["status"] != "SUCCESS":
                last_error = f"Healed collector failed to run: {run_res['error']}"
                continue

            new_records = run_res["records"]
            attempt.records_after = len(new_records)

            new_val_res = OutputValidator.validate_records(
                records=new_records,
                schema=scraper.schema_definition,
                expected_min_count=1,
                historical_avg_count=int(hist_avg) if hist_avg else None
            )

            attempt.validation_result = new_val_res
            attempt.completed_at = datetime.datetime.utcnow()

            if new_val_res["success"]:
                attempt.status = "SUCCESS"

                scraper.status = "HEALTHY"
                scraper.last_success = datetime.datetime.utcnow()
                scraper.last_run = datetime.datetime.utcnow()
                scraper.current_version += 1

                failed_run.status = "HEALED"
                failed_run.recovery_source = "AI_HEAL"
                failed_run.records_count = len(new_records)
                failed_run.validation_status = new_val_res
                failed_run.completed_at = datetime.datetime.utcnow()
                failed_run.duration = (failed_run.completed_at - failed_run.started_at).total_seconds()

                result_filename = f"run_{failed_run.id}_healed.json"
                cls._save_raw_result(result_filename, new_records)
                failed_run.raw_result_reference = result_filename

                new_selectors = BrightDataService.get_selectors(scraper_id=scraper.id)
                new_sv = SelectorVersion(
                    scraper_id=scraper.id,
                    version=scraper.current_version,
                    selectors=new_selectors,
                    success_count=1
                )
                db.add(new_sv)

                BrightDataService.update_scraper_cache(db, scraper.id, new_records, run_res.get("hash"))
                cls.correlate_scrapers_selector_update(db, scraper.id, new_selectors)

                confidence = cls.calculate_confidence_score(new_val_res, hist_avg)
                logger.info(f"Scraper {scraper.name} healed successfully with confidence {confidence}%")

                cls._update_scraper_success_rate(db, scraper_id)
                db.commit()
                return attempt

            last_error = "Post-healing validation failed. Structure is still incorrect."

        # All heal attempts exhausted
        attempt.status = "FAILED"
        attempt.error = last_error or "All healing attempts failed."
        attempt.completed_at = datetime.datetime.utcnow()
        if new_val_res:
            attempt.validation_result = new_val_res

        failed_run.status = "VALIDATION_FAILED"
        failed_run.error = attempt.error
        if new_val_res:
            failed_run.validation_status = new_val_res
        failed_run.completed_at = datetime.datetime.utcnow()
        failed_run.duration = (failed_run.completed_at - failed_run.started_at).total_seconds()

        scraper.status = "ESCALATED"

        cls._update_scraper_success_rate(db, scraper_id)
        db.commit()
        return attempt

    @classmethod
    def correlate_scrapers_selector_update(cls, db: Session, healed_scraper_id: str, new_selectors: Dict[str, str]):
        """
        Cross-scraper correlation:
        Finds other scrapers targeting similar domains or patterns that are currently FAILING or ESCALATED,
        and pre-emptively updates their selectors to the newly healed selectors to restore their health.
        """
        from database.models import Scraper, SelectorVersion
        from urllib.parse import urlparse
        
        healed_scraper = db.query(Scraper).filter(Scraper.id == healed_scraper_id).first()
        if not healed_scraper:
            return
            
        try:
            healed_domain = urlparse(healed_scraper.target_url).netloc
            if not healed_domain:
                return
                
            # Find scrapers with similar domains that are NOT healthy
            correlated_scrapers = db.query(Scraper).filter(
                Scraper.id != healed_scraper_id,
                Scraper.status.in_(["FAILING", "ESCALATED", "VALIDATION_FAILED"])
            ).all()
            
            for cs in correlated_scrapers:
                cs_domain = urlparse(cs.target_url).netloc
                if cs_domain == healed_domain:
                    logger.info(f"Cross-scraper correlation: Pre-emptively healing scraper {cs.id} ('{cs.name}') using selectors from {healed_scraper_id}")
                    
                    # Update selectors locally
                    BrightDataService.set_selectors(new_selectors, scraper_id=cs.id)
                    
                    # Save version
                    cs.current_version += 1
                    new_sv = SelectorVersion(
                        scraper_id=cs.id,
                        version=cs.current_version,
                        selectors=new_selectors,
                        success_count=1
                    )
                    db.add(new_sv)
                    
                    # Update status
                    cs.status = "HEALTHY"
                    db.commit()
        except Exception as e:
            logger.error(f"Error in cross-scraper correlation: {str(e)}")

    @classmethod
    def _save_raw_result(cls, filename: str, data: List[Dict]):
        # Save to local storage folder for runs
        os.makedirs("storage/runs", exist_ok=True)
        filepath = os.path.join("storage/runs", filename)
        with open(filepath, "w") as f:
            json.dump(data, f, indent=2)

    @classmethod
    def _update_scraper_success_rate(cls, db: Session, scraper_id: str):
        scraper = db.query(Scraper).filter(Scraper.id == scraper_id).first()
        if not scraper:
            return

        total_runs = db.query(Run).filter(Run.scraper_id == scraper_id).count()
        if total_runs == 0:
            scraper.success_rate = 100.0
            return

        # SUCCESS and HEALED count as successful outputs in the end-to-end reliability pipeline
        successful_runs = db.query(Run).filter(
            Run.scraper_id == scraper_id,
            Run.status.in_(["SUCCESS", "HEALED"])
        ).count()

        scraper.success_rate = round((successful_runs / total_runs) * 100.0, 1)

File: apps/web/src/services/healer.ts
import fs from 'fs';
import path from 'path';
import { db } from '../db';
import { scrapers, runs, healingAttempts, selectorVersions } from '../db/schema';
import { eq, desc, and, inArray } from 'drizzle-orm';
import { BrightDataService } from './brightdata';
import { FailureDetector } from './detector';
import { OutputValidator } from './validator';

const MAX_HEAL_ATTEMPTS = 2;

export class HealerService {
  static calculateConfidenceScore(validationResult: any | null = null, historicalAvg: number | null = null): number {
    if (!validationResult || !validationResult.success) {
      return 30; // Default failure score
    }

    let score = 70; // Base score for passing validation

    const summary = validationResult.summary || {};
    const fieldsPresence = summary.fields_presence || {};
    const fieldsTypeValid = summary.fields_type_valid || {};
    const totalRecords = summary.total_records || 0;

    // 1. Field completeness bonus (up to 10%)
    const presenceVals = Object.values(fieldsPresence) as number[];
    if (presenceVals.length > 0) {
      const avgPresence = presenceVals.reduce((a, b) => a + b, 0) / presenceVals.length;
      score += Math.floor(avgPresence * 10);
    }

    // 2. Schema validity bonus (up to 10%)
    const typeVals = Object.values(fieldsTypeValid) as number[];
    if (typeVals.length > 0) {
      const avgTypeValid = typeVals.reduce((a, b) => a + b, 0) / typeVals.length;
      score += Math.floor(avgTypeValid * 10);
    }

    // 3. Historical count recovery bonus (up to 10%)
    if (historicalAvg !== null && historicalAvg > 0) {
      const countRatio = Math.min(totalRecords, historicalAvg) / Math.max(totalRecords, historicalAvg);
      score += Math.floor(countRatio * 10);
    } else {
      if (totalRecords > 0) {
        score += 5;
      }
    }

    return Math.min(Math.max(score, 0), 100);
  }

  static async triggerHealingProcess(scraperId: string, failedRunId: string): Promise<any> {
    const scraperList = await db.select().from(scrapers).where(eq(scrapers.id, scraperId)).limit(1);
    const scraper = scraperList[0];

    const runList = await db.select().from(runs).where(eq(runs.id, failedRunId)).limit(1);
    const failedRun = runList[0];

    if (!scraper || !failedRun) {
      console.error(`Cannot heal: Scraper ${scraperId} or Run ${failedRunId} not found.`);
      throw new Error('Scraper or Run not found');
    }

    // 1. Fetch previous successful runs to establish baseline
    const successRuns = await db.select()
      .from(runs)
      .where(and(eq(runs.scraperId, scraperId), eq(runs.status, 'SUCCESS')))
      .orderBy(desc(runs.completedAt))
      .limit(5);

    let histAvg: number | null = null;
    if (successRuns.length > 0) {
      histAvg = successRuns.reduce((sum, r) => sum + r.recordsCount, 0) / successRuns.length;
    }

    // 2. Perform failure detection
    const valStatus = (failedRun.validationStatus as any) || {};
    const failureDescription = FailureDetector.generateDiagnosticSummary(
      scraper.name,
      scraper.collectorId || 'c_mock',
      scraper.schemaDefinition as Record<string, string>,
      valStatus
    );

    // 3. Record Healing Attempt
    const attemptId = crypto.randomUUID();
    await db.insert(healingAttempts).values({
      id: attemptId,
      scraperId: scraper.id,
      runId: failedRun.id,
      status: 'HEALING',
      failureDescription,
      collectorId: scraper.collectorId,
      recordsBefore: failedRun.recordsCount,
      startedAt: new Date(),
    });

    // Update state to HEALING
    await db.update(runs).set({ status: 'HEALING' }).where(eq(runs.id, failedRun.id));
    await db.update(scrapers).set({ status: 'HEALING' }).where(eq(scrapers.id, scraper.id));

    // 3.5. Try Local Selector Version Recovery
    const historicalVersions = await db.select()
      .from(selectorVersions)
      .where(eq(selectorVersions.scraperId, scraperId))
      .orderBy(desc(selectorVersions.successCount), desc(selectorVersions.version));

    let localHealed = false;
    let healedSelectors: any = null;
    let healedVersionObj: any = null;
    let testRunRes: any = null;
    let testValRes: any = null;

    const originalSelectors = { ...BrightDataService.getSelectors(scraper.id) };

    for (const hv of historicalVersions) {
      // Check if selectors match current active selectors
      const hvSelectors = hv.selectors as Record<string, string>;
      const match = Object.keys(hvSelectors).every(k => hvSelectors[k] === originalSelectors[k]) &&
                    Object.keys(originalSelectors).every(k => hvSelectors[k] === originalSelectors[k]);
      if (match) continue;

      console.log(`Attempting local selector recovery using version ${hv.version} for scraper ${scraperId}...`);
      BrightDataService.setSelectors(hvSelectors, scraper.id);

      testRunRes = await BrightDataService.runCollector(
        scraper.collectorId || 'c_mock',
        scraper.targetUrl,
        scraper.schemaDefinition as Record<string, string>,
        scraper.id,
        true // bypass cache
      );

      if (testRunRes.status === 'SUCCESS') {
        testValRes = OutputValidator.validateRecords(
          testRunRes.records,
          scraper.schemaDefinition as Record<string, string>,
          1,
          histAvg ? Math.floor(histAvg) : null
        );

        if (testValRes.success) {
          console.log(`Local selector recovery SUCCESS using version ${hv.version}!`);
          localHealed = true;
          healedSelectors = hvSelectors;
          healedVersionObj = hv;

          // Increment success count
          await db.update(selectorVersions)
            .set({ successCount: hv.successCount + 1 })
            .where(eq(selectorVersions.id, hv.id));
          break;
        }
      }
    }

    if (localHealed && testRunRes && testValRes && healedVersionObj) {
      const now = new Date();
      await db.update(healingAttempts).set({
        status: 'SUCCESS',
        completedAt: now,
        error: null,
        recordsAfter: testRunRes.records.length,
        validationResult: testValRes,
      }).where(eq(healingAttempts.id, attemptId));

      const duration = (now.getTime() - failedRun.startedAt.getTime()) / 1000;
      const resultFilename = `run_${failedRun.id}_healed.json`;
      this.saveRawResult(resultFilename, testRunRes.records);

      await db.update(runs).set({
        status: 'HEALED',
        completedAt: now,
        recordsCount: testRunRes.records.length,
        validationStatus: testValRes,
        recoverySource: 'LOCAL_VERSION_RECOVERY',
        duration,
        rawResultReference: resultFilename,
      }).where(eq(runs.id, failedRun.id));

      await db.update(scrapers).set({
        status: 'HEALTHY',
        currentVersion: healedVersionObj.version,
        lastSuccess: now,
        lastRun: now,
      }).where(eq(scrapers.id, scraper.id));

      // Cache records
      await BrightDataService.updateScraperCache(scraper.id, testRunRes.records, testRunRes.hash || '');

      // Correlate
      await this.correlateScrapersSelectorUpdate(scraper.id, healedSelectors);
      await this.updateScraperSuccessRate(scraper.id);

      const updatedAttempt = await db.select().from(healingAttempts).where(eq(healingAttempts.id, attemptId)).limit(1);
      return updatedAttempt[0];
    }

    // Restore original selectors if local version recovery failed
    BrightDataService.setSelectors(originalSelectors, scraper.id);

    // 4. Fallback to Bright Data AI Healing CLI (with retry loop)
    let lastError: string | null = null;
    let finalAttempt: any = null;

    for (let healAttempt = 1; healAttempt <= MAX_HEAL_ATTEMPTS; healAttempt++) {
      if (healAttempt > 1) {
        console.log(`Retrying AI heal (attempt ${healAttempt}/${MAX_HEAL_ATTEMPTS})...`);
      }

      const healRes = await BrightDataService.healCollector(
        scraper.collectorId || 'c_mock',
        failureDescription,
        scraper.id
      );

      if (healRes.status !== 'SUCCESS') {
        lastError = `Bright Data CLI Healing failed: ${healRes.error}`;
        continue;
      }

      await db.update(healingAttempts).set({ status: 'VALIDATING' }).where(eq(healingAttempts.id, attemptId));

      const runRes = await BrightDataService.runCollector(
        scraper.collectorId || 'c_mock',
        scraper.targetUrl,
        scraper.schemaDefinition as Record<string, string>,
        scraper.id,
        true
      );

      if (runRes.status !== 'SUCCESS') {
        lastError = `Healed collector failed to run: ${runRes.error}`;
        continue;
      }

      const newRecords = runRes.records;
      const newValRes = OutputValidator.validateRecords(
        newRecords,
        scraper.schemaDefinition as Record<string, string>,
        1,
        histAvg ? Math.floor(histAvg) : null
      );

      const now = new Date();
      if (newValRes.success) {
        await db.update(healingAttempts).set({
          status: 'SUCCESS',
          completedAt: now,
          recordsAfter: newRecords.length,
          validationResult: newValRes,
        }).where(eq(healingAttempts.id, attemptId));

        const nextVersion = scraper.currentVersion + 1;
        await db.update(scrapers).set({
          status: 'HEALTHY',
          lastSuccess: now,
          lastRun: now,
          currentVersion: nextVersion,
        }).where(eq(scrapers.id, scraper.id));

        const duration = (now.getTime() - failedRun.startedAt.getTime()) / 1000;
        const resultFilename = `run_${failedRun.id}_healed.json`;
        this.saveRawResult(resultFilename, newRecords);

        await db.update(runs).set({
          status: 'HEALED',
          recoverySource: 'AI_HEAL',
          recordsCount: newRecords.length,
          validationStatus: newValRes,
          completedAt: now,
          duration,
          rawResultReference: resultFilename,
        }).where(eq(runs.id, failedRun.id));

        const newSelectors = BrightDataService.getSelectors(scraper.id);
        await db.insert(selectorVersions).values({
          id: crypto.randomUUID(),
          scraperId: scraper.id,
          version: nextVersion,
          selectors: newSelectors,
          successCount: 1,
          createdAt: now,
        });

        await BrightDataService.updateScraperCache(scraper.id, newRecords, runRes.hash || '');
        await this.correlateScrapersSelectorUpdate(scraper.id, newSelectors);
        await this.updateScraperSuccessRate(scraper.id);

        const updatedAttempt = await db.select().from(healingAttempts).where(eq(healingAttempts.id, attemptId)).limit(1);
        return updatedAttempt[0];
      }

      lastError = 'Post-healing validation failed. Structure is still incorrect.';
      finalAttempt = { newValRes, newRecords, now };
    }

    // All heal attempts exhausted
    const now = new Date();
    const errMsg = lastError || 'All healing attempts failed.';

    await db.update(healingAttempts).set({
      status: 'FAILED',
      completedAt: now,
      error: errMsg,
      validationResult: finalAttempt?.newValRes ?? null,
    }).where(eq(healingAttempts.id, attemptId));

    const duration = (now.getTime() - failedRun.startedAt.getTime()) / 1000;
    await db.update(runs).set({
      status: 'VALIDATION_FAILED',
      error: errMsg,
      validationStatus: finalAttempt?.newValRes ?? null,
      completedAt: now,
      duration,
    }).where(eq(runs.id, failedRun.id));

    await db.update(scrapers).set({ status: 'ESCALATED' }).where(eq(scrapers.id, scraper.id));
    await this.updateScraperSuccessRate(scraper.id);

    const finalAttemptRow = await db.select().from(healingAttempts).where(eq(healingAttempts.id, attemptId)).limit(1);
    return finalAttemptRow[0];
  }

  static async correlateScrapersSelectorUpdate(healedScraperId: string, newSelectors: Record<string, string>) {
    const scraperList = await db.select().from(scrapers).where(eq(scrapers.id, healedScraperId)).limit(1);
    const healedScraper = scraperList[0];
    if (!healedScraper) return;

    try {
      const healedDomain = new URL(healedScraper.targetUrl).hostname;
      if (!healedDomain) return;

      const correlatedScrapers = await db.select()
        .from(scrapers)
        .where(and(
          inArray(scrapers.status, ['FAILING', 'ESCALATED', 'VALIDATION_FAILED']),
          // We filter matching domains dynamically below
        ));

      for (const cs of correlatedScrapers) {
        let csDomain = '';
        try {
          csDomain = new URL(cs.targetUrl).hostname;
        } catch {}

        if (csDomain === healedDomain && cs.id !== healedScraperId) {
          console.log(`Cross-scraper correlation: Pre-emptively healing scraper ${cs.id} ('${cs.name}') using selectors from ${healedScraperId}`);

          // Update selector mapping in memory
          BrightDataService.setSelectors(newSelectors, cs.id);

          const nextVersion = cs.currentVersion + 1;
          await db.update(scrapers).set({
            currentVersion: nextVersion,
            status: 'HEALTHY',
            updatedAt: new Date(),
          }).where(eq(scrapers.id, cs.id));

          // Save new version
          await db.insert(selectorVersions).values({
            id: crypto.randomUUID(),
            scraperId: cs.id,
            version: nextVersion,
            selectors: newSelectors,
            successCount: 1,
          });
        }
      }
    } catch (e) {
      console.error('Error in cross-scraper correlation:', e);
    }
  }

  private static saveRawResult(filename: string, data: any[]) {
    try {
      const dir = path.join(process.cwd(), 'storage', 'runs');
      if (!fs.existsSync(dir)) {
        fs.mkdirSync(dir, { recursive: true });
      }
      fs.writeFileSync(path.join(dir, filename), JSON.stringify(data, null, 2));
    } catch (e) {
      console.error('Error saving raw run results:', e);
    }
  }

  static async updateScraperSuccessRate(scraperId: string) {
    try {
      const runList = await db.select().from(runs).where(eq(runs.scraperId, scraperId));
      const totalRuns = runList.length;

      if (totalRuns === 0) {
        await db.update(scrapers).set({ successRate: 100.0 }).where(eq(scrapers.id, scraperId));
        return;
      }

      const successfulRuns = runList.filter(r => ['SUCCESS', 'HEALED'].includes(r.status)).length;
      const successRate = parseFloat(((successfulRuns / totalRuns) * 100.0).toFixed(1));

      await db.update(scrapers).set({ successRate }).where(eq(scrapers.id, scraperId));
    } catch (e) {
      console.error('Error updating scraper success rate:', e);
    }
  }
}
